src/common/download.h

```diff
@@ -32,5 +32,11 @@
 
 namespace tools
 {
-  bool download(const std::string &path, const std::string &url);
+  typedef void *download_async_handle;
```

Comment:
Ah sorry for more comments, but I just got a better idea. `future`s don't really have a good way to cancel. But:
```c++
class download_async_handle {
  struct impl;
  std::shared_ptr<impl> instance;
public:
  download_async_handle() : instance(nullptr) {}
  explicit download_async_handle(std::shared_ptr<impl> src)
    : instance(std::move(src)) {}

  download_async_handle(const download_async_handle&) = delete;
  download_async_handle(download_async_handle&&) = default;

  download_async_handle& operator=(const download_async_handle&) = delete;
  download_async_handle& operator=(download_async_handle&&) = default;

  ~download_async_handle() noexcept { cancel_async(); }

  void wait(); // waits and joins
  void cancel(); // tells the thread to stop and joins
  void cancel_async() noexcept; // tells the thread to stop, but does not join
};
download_async_handle download_async(...);
```
Since the class is non-copyable, a download thread will only be associated to one `download_async_handle`. Which means the destructor can tell the server to stop (probably through an atomic), and let the thread exit _eventually_. The `shared_ptr<>` cleans up the memory to whoever finishes last.